### PR TITLE
Add get_byte_slice/byte_slice

### LINKED ISSUE #49: Add slice_byte() / get_slice_byte()
It would be a nice optimization in Helix since we receive byte ranges from tree-sitter then convert them into char indices, but then ropey converts back to bytes and uses byte calculations internally.

### PR DESCRIPTION
**Pull request by archseer**
Should probably state in the docs that this is marginally faster than `slice()` if you already have byte indices but that `slice()` should be preferred for safety (index is always aligned).

Fixes #49

### CODE REVIEW

File: src/slice.rs [hunk 1]
```diff
@@ -1229,6 +1370,53 @@ impl<'a> RopeSlice<'a> {
...
+                RopeSlice(RSEnum::Full {
+                    node, start_info, ..
+                }) => Some(RopeSlice::new_with_byte_range(
+                    node,
+                    start_info.bytes as usize + start,
+                    start_info.bytes as usize + end,
+                )),
+                RopeSlice(RSEnum::Light { text, .. }) => {
+                    let start_byte = start;
+                    let end_byte = end;
```
Comment (archseer) [inline src/slice.rs:1405]: I kept the name here but this could be cleaned up by renaming the variables further up

File: src/slice.rs [hunk 2]
```diff
@@ -127,6 +127,87 @@ impl<'a> RopeSlice<'a> {
...
+                // simpler lightweight slice then.
+                Node::Leaf(ref text) => {
+                    let start_byte = n_start;
+                    let end_byte = n_end;
+                    return RopeSlice(RSEnum::Light {
+                        text: &text[start_byte..end_byte],
+                        char_count: count_chars(&text[start_byte..end_byte]) as Count,
+                        utf16_surrogate_count: count_utf16_surrogates(&text[start_byte..end_byte])
+                            as Count,
+                        line_break_count: count_line_breaks(&text[start_byte..end_byte]) as Count,
```
Comment (archseer) [inline src/slice.rs:183]: I kept the pattern here from `new_with_range` but the `&text[start_byte..end_byte]` can probably be extracted out

**Comment by cessen** [comment]
Sorry for taking so long to get to this!

The code mostly looks good to me.  However, at the moment this will always panic when trying to slice at a non-char boundary.  And this is a problem for a couple of reasons:

1. It's possible for `get_slice_byte()` to panic, whereas the intention of the `get_*()` methods is that they never panic, they just return `None`.
2.   Ideally I'd like panic error messages to be more informative, and also specific to the actual public method called.  With the code as-is, the panic error message will just be a generic `&str` slicing error message.

Nevertheless, I'm fine merging this as-is and cleaning that up myself when I (eventually) get around to it.  But if you have the time and would like to tackle that yourself, let me know, and I'll hold off on merging.

**Comment by cessen** [comment]
On second thought, I think I'd rather tackle the clean up myself, because I'm realizing that to do it "properly" I'll want to rework some other things in Ropey as well.  So I'll merge this as-is, and work on that myself when I get the time.

**Pull request merged**